File: Data_Structure_using_JavaScript/SinglyCircularLL.js
class node
{
    data
    next

    constructor(no)
    {
        this.data = no
        this.next = null
    }
}

class SinglyCLL
{
    #first
    #last
    #iCount

    constructor()
    {
        this.#first = null
        this.#last = null
        this.#iCount = 0
    }

    InsertFirst(iNo)
    {
        let newn = null

        newn = new node(iNo)

        if((null == this.#first) && (null == this.#last))
        {
            this.#first = newn
            this.#last = newn
        }
        else
        {
            newn.next = this.#first
            this.#first = newn
        }
        this.#last.next = this.#first
        this.#iCount++
    }

    InsertLast(iNo)
    {
        let newn = null

        newn = new node(iNo)

        if((null == this.#first) && (null == this.#last))
        {
            this.#first = newn
            this.#last = newn
        }
        else
        {
            this.#last.next = newn
            this.#last = newn
        }
        this.#last.next = this.#first
        this.#iCount++
    }

    InsertAtPos(iNo, iPos)
    {
        let newn = null
        let temp = null
        let iCnt = 0

        if((iPos < 1) || (iPos > this.#iCount+1))
        {
            console.log("Invalid Position")
            return
        }

        if(iPos == 1)
        {
            this.InsertFirst(iNo)
        }
        else if(iPos == this.#iCount+1) 
        {
            this.InsertLast(iNo)   
        }
        else
        {
            newn = new node(iNo)

            temp = this.#first
            for(iCnt = 1;iCnt < iPos-1;iCnt++)
            {
                temp = temp.next
            }
            newn.next = temp.next
            temp.next = newn
            this.#iCount++
        }
    }

    DeleteFirst()
    {
        if((null == this.#first) && (null == this.#last))
        {
            console.log("Linked List is empty")
            return
        }
        else if(this.#first == this.#last)
        {
            this.#first = null
            this.#last = null
        }
        else
        {
            this.#first = this.#first.next
            this.#last.next = this.#first
        }
        this.#iCount--
    }

    DeleteLast()
    {
        let temp = null

        if((null == this.#first) && (null == this.#last))
        {
            console.log("Linked List is empty")
            return
        }
        else if(this.#first == this.#last)
        {
            this.#first = null
            this.#last = null
        }
        else
        {
            temp = this.#first
            while(temp.next != this.#last)
            {
                temp = temp.next
            }
            this.#last = temp
            this.#last.next = this.#first
        }
        this.#iCount--
    }

    DeleteAtPos(iPos)
    {
        let temp = null
        let iCnt = 0

        if((iPos < 1) || (iPos > this.#iCount))
        {
            console.log("Invalid Position")
            return
        }

        if(iPos == 1)
        {
            this.DeleteFirst()
        }
        else if(iPos == this.#iCount) 
        {
            this.DeleteLast()
        }
        else
        {
            temp = this.#first
            for(iCnt = 1;iCnt < iPos-1;iCnt++)
            {
                temp = temp.next
            }

            temp.next = temp.next.next
            this.#iCount--
        }
    }

    Display()
    {
        let temp = null
        temp = this.#first
        do
        {
            process.stdout.write("| "+temp.data+" |->")
            temp = temp.next
        }while(temp != this.#last.next)
        console.log()
    }

    Count()
    {
        return this.#iCount
    }
}

console.log()
console.log("---------------------------------- For Number ----------------------------------------")
console.log()

const sobj = new SinglyCLL()
let iRet = 0

sobj.InsertFirst(51)
sobj.InsertFirst(21)
sobj.InsertFirst(11)

sobj.InsertLast(101)
sobj.InsertLast(111)
sobj.InsertLast(121)
sobj.InsertLast(151)

sobj.Display()

iRet = sobj.Count()

console.log("Number of nodes in LL is : "+iRet)

sobj.InsertAtPos(105,5)

sobj.Display()

iRet = sobj.Count()

console.log("Number of nodes in LL is : "+iRet)

sobj.DeleteAtPos(5)

sobj.Display()

iRet = sobj.Count()

console.log("Number of nodes in LL is : "+iRet)

sobj.DeleteFirst()

sobj.Display()

iRet = sobj.Count()

console.log("Number of nodes in LL is : "+iRet)

sobj.DeleteLast()

sobj.Display()

iRet = sobj.Count()

console.log("Number of nodes in LL is : "+iRet)

/////////////////////////////////////////////////////////////////////////
console.log()
console.log("---------------------------------- For Bigint ----------------------------------------")
console.log()

const sobj2 = new SinglyCLL()

sobj2.InsertFirst(516354653165n)
sobj2.InsertFirst(21876496896854n)
sobj2.InsertFirst(11874685466666n)

sobj2.InsertLast(101643618543166635n)
sobj2.InsertLast(111963959683565266n)
sobj2.InsertLast(121689365163656565n)
sobj2.InsertLast(1513638546354636565n)

sobj2.Display()

iRet = sobj2.Count()

console.log("Number of nodes in LL is : "+iRet)

sobj2.InsertAtPos(105,5)

sobj2.Display()

iRet = sobj2.Count()

console.log("Number of nodes in LL is : "+iRet)

sobj2.DeleteAtPos(5)

sobj2.Display()

iRet = sobj2.Count()

console.log("Number of nodes in LL is : "+iRet)

sobj2.DeleteFirst()

sobj2.Display()

iRet = sobj2.Count()

console.log("Number of nodes in LL is : "+iRet)

sobj2.DeleteLast()

sobj2.Display()

iRet = sobj2.Count()

console.log("Number of nodes in LL is : "+iRet)

File: Data_Structure_using_C_Programming/SinglyCircularLL.c
// Singly Circular Linked List

#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node* next;
};

typedef struct node NODE;
typedef struct node* PNODE;
typedef struct node** PPNODE;

void InsertFirst(PPNODE first, PPNODE last, int iNo)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));
    newn->data = iNo;
    newn->next = NULL;

    if((NULL == *first) && (NULL == *last))
    {
        *first = newn;
        *last = newn;
        (*last)->next = *first;
    }
    else
    {
        newn->next = *first;
        *first = newn;
        (*last)->next = *first;
    }
}

void InsertLast(PPNODE first, PPNODE last, int iNo)
{
    PNODE newn = NULL, temp = NULL;

    newn = (PNODE)malloc(sizeof(NODE));
    newn->data = iNo;
    newn->next = NULL;

    if((NULL == *first) && (NULL == *last))
    {
        *first = newn;
        *last = newn;
        (*last)->next = *first;
    }
    else
    {   
        (*last)->next = newn;
        *last = newn;
        (*last)->next = *first;
    }
}

void DeleteFirst(PPNODE first, PPNODE last)
{
    if((NULL == *first) || (NULL == *last))
    {
        return;
    }
    else if(*first == *last)
    {
        free(*first);
        *first = NULL;
        *last = NULL;
    }
    else
    {
        *first = (*first)->next;
        free((*last)->next);
        (*last)->next = *first;
    }
}

void DeleteLast(PPNODE first, PPNODE last)
{
    PNODE temp = NULL;

    if((NULL == *first) || (NULL == *last))
    {
        return;
    }
    else if(*first == *last)
    {
        free(*first);
        *first = NULL;
        *last = NULL;
    }
    else
    {
        temp = *first;

        while(temp->next != *last)
        {
            temp = temp->next;
        }
        
        *last = temp;
        free(temp->next);
        (*last)->next = *first;
    }
}

void Display(PNODE first, PNODE last)
{
    if((first == NULL) && (last == NULL))
    {
        return;
    }

    printf("->");
    do
    {
        printf("| %d |->",first->data);
        first = first->next;
    } while(first != last->next);
    printf("\n");
}

int Count(PNODE first, PNODE last)
{
    int iCount = 0;

    do
    {
        iCount++;
        first = first->next;
    } while(first != last->next);

    return iCount;
}

void InsertAtPos(PPNODE first, PPNODE last, int iNo, int iPos)
{
    int iCount = 0, iCnt = 0;
    iCount = Count(*first,*last);
    PNODE newn = NULL , temp = NULL;

    if((iPos < 1) || (iPos > iCount+1))
    {
        printf("Invalid position\n");
        return;
    }
    
    if(iPos == 1)
    {
        InsertFirst(first,last,iNo);
    }
    else if(iPos == iCount+1)
    {
        InsertLast(first,last,iNo);
    }
    else
    {
        temp = *first;

        for(iCnt = 1;iCnt < iPos-1;iCnt++)
        {
            temp = temp->next;
        }
        
        newn = (PNODE)malloc(sizeof(NODE));
        newn->data = iNo;
        newn->next = NULL;

        newn->next = temp->next;
        temp->next = newn;
    }
}

void DeleteAtPos(PPNODE first, PPNODE last, int iPos)
{
    int iCount = 0, iCnt = 0;
    iCount = Count(*first,*last);
    PNODE target = NULL , temp = NULL;

    if((iPos < 1) || (iPos > iCount))
    {
        printf("Invalid position\n");
        return;
    }
    
    if(iPos == 1)
    {
        DeleteFirst(first,last);
    }
    else if(iPos == iCount)
    {
        DeleteLast(first,last);
    }
    else
    {
        temp = *first;

        for(iCnt = 1;iCnt < iPos-1;iCnt++)
        {
            temp = temp->next;
        }

        target = temp->next;
        temp->next = target->next;
        free(target);
    }
}


int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;
    int iRet = 0;

    InsertFirst(&head,&tail,51);
    InsertFirst(&head,&tail,21);
    InsertFirst(&head,&tail,11);

    InsertLast(&head,&tail,101);
    InsertLast(&head,&tail,111);
    InsertLast(&head,&tail,121);

    Display(head,tail);

    iRet = Count(head,tail);

    printf("Number of elements in Singly Circular LinkedList is : %d\n",iRet);

    InsertAtPos(&head,&tail,105,-1);

    Display(head,tail);

    iRet = Count(head,tail);

    printf("Number of elements in Singly Circular LinkedList is : %d\n",iRet);

    DeleteAtPos(&head,&tail,5);

    Display(head,tail);

    iRet = Count(head,tail);

    printf("Number of elements in Singly Circular LinkedList is : %d\n",iRet);

    return 0;
}